rspace/src/main/scala/coop/rchain/rspace/serializers/ScodecSerialize.scala

```diff
@@ -0,0 +1,345 @@
+package coop.rchain.rspace.serializers
+
+import coop.rchain.rspace.hashing.Blake2b256Hash
+import coop.rchain.rspace.history.PointerBlock.length
+import coop.rchain.rspace.history._
+import coop.rchain.rspace.internal.{Datum, WaitingContinuation}
+import coop.rchain.rspace.trace.{COMM, Consume, Event, Produce}
+import coop.rchain.rspace.util
+import coop.rchain.scodec.codecs.seqOfN
+import coop.rchain.shared.Serialize
+import coop.rchain.shared.Serialize._
+import scodec.bits.ByteVector
+import scodec.codecs.{bool, discriminated, int32, provide, uint, uint2, uint8, vectorOfN}
+import scodec.{Attempt, Codec}
+
+import scala.collection.SortedSet
+import scala.collection.concurrent.TrieMap
+
+/**
+  * This file represents RSpace serializers based on scodec library.
+  *
+  * With [[Serialize]] interface RSpace accepts parametrized serializers
+  * for the main types (channel, pattern, attributes, continuation).
+  *
+  * In the current implementation RSpace internally also uses concrete scodec
+  * serializers for additional types and collections.
+  */
+object ScodecSerialize {
+
+  /*
+   * Datum serializer
+   */
+
+  def encodeDatums[A](datums: Seq[Datum[A]])(implicit sa: Serialize[A]): ByteVector = {
+    val codec = serializeToCodecDatumMemo(sa)
+
+    encodeSortedSeq[Datum[A]](datums, codec)
+  }
+
+  def decodeDatums[A](bytes: ByteVector)(implicit sa: Serialize[A]): Seq[Datum[A]] =
+    decodeDatumsProj[A, Datum[A]](bytes)((d, _) => d)
+
+  def decodeDatumsProj[A, R](
```

Comment:
I am curious what is `Proj` in full?

Reply:
Yes, it's a projection done by function which can modify raw data and produce result as new object. If it's just return all values it would allocate memory for collection of all raw data in cases where only partial data is needed.